// @node-red/ignite/database-operations/models/igniteMetadeta.js
import SequelizeContext from './sequelizeContext';

class IgniteMetadeta extends SequelizeContext {

    constructor() {
        super();
        this._dataType = this.sequelizeDataType;
    }

    get igniteMetadetaSchema() {
        return this.sequelizeConnection.define('ignitemetadata', {
            id: {
                type: this._dataType.INTEGER,
                autoIncrement: true
            },
            metadataname: {
                type: this._dataType.STRING,
                primaryKey: true
            },
            folderdetails: {
                type: this._dataType.TEXT
            },
            extractfromsubfolders: {
                type: this._dataType.BOOLEAN,
                defaultValue: false
            },
            created_date: {
                type: this._dataType.DATE,
                defaultValue: this._sequelizeConnection.literal('CURRENT_TIMESTAMP')
            },
            updated_date: {
                type: this._dataType.DATE,
                defaultValue: this._sequelizeConnection.literal('CURRENT_TIMESTAMP')
            }
        }, {
            timestamps: false,
            freezeTableName: true
        });
    }
}

export default IgniteMetadeta;
