// @node-red/ignite/database-operations/models/projectDetails.js
import SequelizeContext from './sequelizeContext';

class ProjectDetails extends SequelizeContext {

    constructor() {
        super();
        this._dataType = this.sequelizeDataType;
    }

    get projectDetailsSchema() {
        return this.sequelizeConnection.define('projectdetails', {
            id: {
                type: this._dataType.INTEGER,
                autoIncrement: true
            },
            clientid: {
                type: this._dataType.UUID,
                primaryKey: true
            },
            projectsfolderdetails: {
                type: this._dataType.TEXT
            },
            sshfolderdetails: {
                type: this._dataType.TEXT
            },
            created_date: {
                type: this._dataType.DATE,
                defaultValue: this._sequelizeConnection.literal('CURRENT_TIMESTAMP')
            },
            updated_date: {
                type: this._dataType.DATE,
                defaultValue: this._sequelizeConnection.literal('CURRENT_TIMESTAMP')
            }
        }, {
            timestamps: false
        });
    }
}

export default ProjectDetails;
